packages/headless/src/ssr-next/common/builders/controller-builder.ts

```diff
@@ -0,0 +1,60 @@
+import type {CoreEngine, CoreEngineNext} from '../../../app/engine.js';
+import type {Controller} from '../../../controllers/controller/headless-controller.js';
+import type {SSRCommerceEngine} from '../../commerce/factories/build-factory.js';
+import type {
+  BaseControllerDefinitionWithoutProps,
+  BaseControllerDefinitionWithProps,
+} from '../types/controllers.js';
+
+/**
+ * Builder class for creating controllers from definitions
+ */
+export class ControllerBuilder<
+  TDefinition extends Partial<
+    BaseControllerDefinitionWithoutProps<TEngine, TController>
+  > &
+    Partial<BaseControllerDefinitionWithProps<TEngine, TController, TProps>>,
+  TEngine extends CoreEngine | CoreEngineNext | SSRCommerceEngine,
+  TProps,
+  TController extends Controller = Controller,
+> {
+  private definition: TDefinition;
+  private engine: TEngine;
+  private props?: TProps;
+  private additionalArgs: unknown[] = [];
+
+  constructor(definition: TDefinition, engine: TEngine, props?: TProps) {
+    this.definition = definition;
+    this.engine = engine;
+    this.props = props;
+  }
+
+  public setAdditionalArgs(additionalArgs: unknown[] = []): this {
+    this.additionalArgs = additionalArgs;
+    return this;
+  }
+
+  public build(): TController {
+    if ('build' in this.definition) {
+      return this.buildWithoutProps();
+    } else {
```

Comment:
Since `build` and `buildWithProps` both end up being optional in `TDefinition`, should we also cover the (admittedly theoretical) case where neither is present in `this.definition` (I guess we'd just throw an error)? 😅

Reply:
Yes very good point.
This could happen if someone create a custom SSR  component (in javascript without respecting the typings) and forgets to add the `build` or `buildWithProps` methods.